src/storage/file_storage.rs

```diff
@@ -0,0 +1,176 @@
+use std::fs::{self, File};
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::str::FromStr;
+use std::time::SystemTime;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    path::{Path, PathBuf},
+};
+
+use crate::id::ResourceId;
+use crate::{ArklibError, Result};
+
+const LOG_PREFIX: &str = "[file-storage]";
+const STORAGE_VERSION: i32 = 2;
+const STORAGE_VERSION_PREFIX: &str = "version ";
+const KEY_VALUE_SEPARATOR: char = ':';
+
+pub struct FileStorage {
+    log_prefix: String,
+    label: String,
+    path: PathBuf,
+    timestamp: SystemTime,
+}
+
+impl FileStorage {
+    /// Create a new file storage with a diagnostic label and file path
+    pub fn new(label: String, path: &Path) -> Self {
+        Self {
+            label: label.clone(),
+            log_prefix: format!("{} {}", LOG_PREFIX, label).to_string(),
+            path: PathBuf::from(path),
+            timestamp: SystemTime::now(),
+        }
+    }
+
+    /// Read data from disk
+    ///
+    /// Data is read as a key value pairs separated by a symbol and stored
+    /// in a [HashMap] with [ResourceId] key and a generic V value. A handler
+    /// is called on the data after reading it.
+    pub fn read_from_disk<V>(
+        &self,
+        handle: impl FnOnce(HashMap<ResourceId, V>),
+    ) -> Result<()>
+    where
+        V: FromStr,
+        ArklibError: From<<V as FromStr>::Err>,
+    {
+        let new_timestamp = fs::metadata(&self.path)?.modified()?;
+        log::info!(
+            "timestamp of storage file {:?} is {:?}",
+            self.path,
+            self.timestamp
+        );
+
+        if self.timestamp >= new_timestamp {
+            return Ok(());
+        }
+
+        log::info!("the file was modified externally, merging");
+
+        let file = fs::File::open(&self.path)?;
+        let reader = BufReader::new(file);
+        let mut lines = reader.lines();
+
+        match lines.next() {
+            Some(Ok(header)) => {
+                self.verify_version(&header)?;
+
+                let mut value_by_id = HashMap::new();
+                for line in lines {
+                    let line = line?;
+                    if line.is_empty() {
+                        continue;
+                    }
+
+                    let parts: Vec<&str> =
+                        line.split(KEY_VALUE_SEPARATOR).collect();
+                    let id = ResourceId::from_str(parts[0])?;
+                    let value = V::from_str(parts[1])?;
+
+                    // TODO: neutral value check when introducing monoids
```

Comment:
TODOs should be resolved before merging.

Reply:
https://github.com/ARK-Builders/arklib-android/blob/377734870453889f3e9a8821bb1cac3f7f299ab8/lib/src/main/java/dev/arkbuilders/arklib/data/metadata/MetadataStorage.kt#L18

`isNeutral` is defined here in arklib-android repo. I'm not sure how it's related to file storage and why it's a constant.

Reply:
This one in `MetadataStorage` is only one of the cases. It's defined as `false` because monoid-based merge strategy isn't used there. See `MonoidIsNotUsed` in the type header. It isn't used because it's cache and we always want to use it locally. That's one of those moments when you implement a universal approach and then one thing doesn't fit nicely into it...

In storages where monoid-based merge-strategy is used, `isNeutral` means same as `== 0`, or `== emptySet()` etc. Depends on what kind of monoid is used.

Ideally, we want to use something more flexible than Monoid which can only sum things up. Implementing CRDT would be great, but we can leave it for next PRs.

Reply:
Ok, I'll remove the TODOs, We'll take this up as a separate issue when we look into merging data.